src/main/java/com/google/firebase/internal/RetryConfig.java

```diff
@@ -0,0 +1,160 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.util.BackOff;
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.api.client.util.Sleeper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Configures when and how HTTP requests should be retried.
+ */
+public final class RetryConfig {
+
+  private static final int INITIAL_INTERVAL_MILLIS = 500;
+
+  private final List<Integer> retryStatusCodes;
+  private final int maxRetries;
+  private final Sleeper sleeper;
+  private final ExponentialBackOff.Builder backOffBuilder;
+
+  private RetryConfig(Builder builder) {
+    if (builder.retryStatusCodes != null) {
+      this.retryStatusCodes = ImmutableList.copyOf(builder.retryStatusCodes);
+    } else {
+      this.retryStatusCodes = ImmutableList.of();
+    }
+
+    checkArgument(builder.maxRetries >= 0, "maxRetries must not be negative");
+    this.maxRetries = builder.maxRetries;
+    this.sleeper = checkNotNull(builder.sleeper);
+    this.backOffBuilder = new ExponentialBackOff.Builder()
+        .setInitialIntervalMillis(INITIAL_INTERVAL_MILLIS)
+        .setMaxIntervalMillis(builder.maxIntervalMillis)
+        .setMultiplier(builder.backOffMultiplier)
+        .setRandomizationFactor(0);
+
+    // Force validation of arguments by building the BackOff object
+    this.backOffBuilder.build();
+  }
+
+  List<Integer> getRetryStatusCodes() {
+    return retryStatusCodes;
+  }
+
+  int getMaxRetries() {
+    return maxRetries;
+  }
+
+  int getMaxIntervalMillis() {
+    return backOffBuilder.getMaxIntervalMillis();
+  }
+
+  double getBackOffMultiplier() {
+    return backOffBuilder.getMultiplier();
+  }
+
+  Sleeper getSleeper() {
+    return sleeper;
+  }
+
+  BackOff newBackOff() {
+    return backOffBuilder.build();
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static final class Builder {
+
+    private List<Integer> retryStatusCodes;
+    private int maxRetries;
+    private int maxIntervalMillis = (int) TimeUnit.MINUTES.toMillis(2);
+    private double backOffMultiplier = 2.0;
+    private Sleeper sleeper = Sleeper.DEFAULT;
+
+    private Builder() { }
+
+    /**
+     * Sets a list of HTTP status codes that should be retried. If null or empty, HTTP requests
+     * will not be retried as long as they result in some HTTP response message. I/O errors
+     * will still be retried.
```

Comment:
We have had some long discussion with the Firestore backend team about retrying requests that were dropped and never responded to. Unless we no for certain that no byte made it across the wire, it was decided that we cannot retry. Since you are essentially building a retry framework for a wide variety of consumers, I would advise that we don't retry I/O errors by default.

Reply:
In Python and Go we have enabled this, but I can see why this can be a bad idea in some cases. Added an option to make this configurable via `RetryConfig`.